File: app/src/main/java/xyz/lrhm/komakdast/core/util/legacy/ImageManager.java
package xyz.lrhm.komakdast.core.util.legacy;

import android.app.ActivityManager;
import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.ColorMatrix;
import android.graphics.ColorMatrixColorFilter;
import android.graphics.Paint;
import android.graphics.Rect;
import android.os.Build;
import android.util.LruCache;
import android.widget.ImageView;

import java.io.InputStream;

public class ImageManager {
    private Context context;
    private LengthManager lengthManager;
    private int memoryClass;

    private static boolean cacheInited = false;
    private static Object getLock = new Object();
    private static ImageManager instance;

    public static ImageManager getInstance(Context context) {
        synchronized (getLock) {
            if (instance == null)
                instance = new ImageManager(context);
            return instance;
        }
    }

    private ImageManager(Context context) {


        initCache(context);
        this.context = context;
        lengthManager = new LengthManager(context);

    }

    public int getMemoryClass() {
        return memoryClass;
    }

    private void initCache(Context context) {

        if (cacheInited)
            return;
        cacheInited = true;

        ActivityManager am = (ActivityManager) context
                .getSystemService(Context.ACTIVITY_SERVICE);
        memoryClass = am.getMemoryClass();

        int max = (int) ((memoryClass * 1024 * 1024) * ((memoryClass > 100) ? 0.80
                : (0.666))); // more than 50%
        cache = new LruCache<ImageKey, Bitmap>(max) {
            @Override
            protected int sizeOf(ImageKey key, Bitmap value) {
                if (Build.VERSION.SDK_INT >= 12) {
                    return value.getByteCount();
                } else {

                    return value.getRowBytes() * value.getHeight();
                }
            }
        };

    }

    static LruCache<ImageKey, Bitmap> cache;

    public Bitmap loadImageFromResource(int resourceId, int outWidth, int outHeight, ScalingLogic scalingLogic, Bitmap.Config config) {
        if (outWidth == -1) outWidth = lengthManager.getWidthWithFixedHeight(resourceId, outHeight);
        if (outHeight == -1)
            outHeight = lengthManager.getHeightWithFixedWidth(resourceId, outWidth);

        ImageKey key = new ImageKey(resourceId, outWidth, outHeight);

        Bitmap him = cache.get(key);
        if (him != null && !him.isRecycled())
            return him;

        System.gc();

        Bitmap scaledBitmap;
        Bitmap unscaledBitmap;

        unscaledBitmap = decodeFile(resourceId, outWidth, outHeight, scalingLogic, context.getResources());

        scaledBitmap = createScaledBitmap(unscaledBitmap, outWidth, outHeight, scalingLogic, config);
        if (!unscaledBitmap.isRecycled()) unscaledBitmap.recycle();

        cache.put(key, scaledBitmap);
        return scaledBitmap;
    }


    public Bitmap loadImageFromFilse(String path, int outWidth, int outHeight, ScalingLogic scalingLogic) {


        ImageKey key = new ImageKey(path, outWidth, outHeight);

        Bitmap him = cache.get(key);
        if (him != null && !him.isRecycled())
            return him;

        System.gc();

        Bitmap scaledBitmap;
        Bitmap unscaledBitmap;

        unscaledBitmap = decodeFile(path, outWidth, outHeight, scalingLogic, context.getResources());
        scaledBitmap = createScaledBitmap(unscaledBitmap, outWidth, outHeight, scalingLogic, Bitmap.Config.ARGB_8888);
        if (!unscaledBitmap.isRecycled()) unscaledBitmap.recycle();

        cache.put(key, scaledBitmap);
        return scaledBitmap;
    }


    public Bitmap loadImageFromResourceNoCache(int resourceId, int outWidth, int outHeight, ScalingLogic scalingLogic, Bitmap.Config config) {

        return loadImageFromResource(resourceId, outWidth, outHeight, scalingLogic, config);

    }

    public Bitmap loadImageFromResourceNoCache(int resourceId, int outWidth, int outHeight, ScalingLogic scalingLogic) {

        return loadImageFromResource(resourceId, outWidth, outHeight, scalingLogic, Bitmap.Config.ARGB_8888);

    }

    public Bitmap loadImageFromResource(int resourceId, int outWidth, int outHeight, ScalingLogic scalingLogic) {

        return loadImageFromResource(resourceId, outWidth, outHeight, scalingLogic, Bitmap.Config.ARGB_8888);

    }

    public Bitmap loadImageFromResource(int resourceId, int outWidth, int outHeight) {
        return loadImageFromResource(resourceId, outWidth, outHeight, ScalingLogic.CROP, Bitmap.Config.ARGB_8888);
    }


    public Bitmap loadImageFromResource(int resourceId, int outWidth, int outHeight, Bitmap.Config config) {
        return loadImageFromResource(resourceId, outWidth, outHeight, ScalingLogic.CROP, config);
    }


    public Bitmap loadImageFromInputStream(InputStream inputStream, int outWidth, int outHeight) {
        System.gc();

        Bitmap scaledBitmap;
        Bitmap unscaledBitmap;

        if (inputStream == null)
            throw new IllegalStateException("null InputStream!");

        unscaledBitmap = decodeInputStream(inputStream);

        if (outHeight == -1)
            outHeight = unscaledBitmap.getHeight() * outWidth / unscaledBitmap.getWidth();

        if (outWidth == -1)
            outWidth = unscaledBitmap.getWidth() * outHeight / unscaledBitmap.getHeight();

        scaledBitmap = createScaledBitmap(unscaledBitmap, outWidth, outHeight, ScalingLogic.CROP, Bitmap.Config.ARGB_8888);

        if (!unscaledBitmap.isRecycled()) unscaledBitmap.recycle();

        return scaledBitmap;
    }

    public enum ScalingLogic {FIT, CROP, ALL_TOP}

    public Rect calculateSrcRect(int srcWidth, int srcHeight, int dstWidth, int dstHeight, ScalingLogic scalingLogic) {
        switch (scalingLogic) {
            case CROP:
                final float srcAspect = (float) srcWidth / (float) srcHeight;
                final float dstAspect = (float) dstWidth / (float) dstHeight;
                if (srcAspect > dstAspect) {
                    final int srcRectWidth = (int) (srcHeight * dstAspect);
                    final int srcRectLeft = (srcWidth - srcRectWidth) / 2;
                    return new Rect(srcRectLeft, 0, srcRectLeft + srcRectWidth, srcHeight);
                } else {
                    final int srcRectHeight = (int) (srcWidth / dstAspect);
                    final int scrRectTop = (srcHeight - srcRectHeight) / 2;
                    return new Rect(0, scrRectTop, srcWidth, scrRectTop + srcRectHeight);
                }
            case FIT:
                return new Rect(0, 0, srcWidth, srcHeight);
            case ALL_TOP:
                return new Rect(0, 0, srcWidth, Math.min(srcHeight, dstHeight * srcWidth / dstWidth));
            default:
                return null;
        }
    }

    public Rect calculateDstRect(int srcWidth, int srcHeight, int dstWidth, int dstHeight, ScalingLogic scalingLogic) {
        switch (scalingLogic) {
            case FIT:
                final float srcAspect = (float) srcWidth / (float) srcHeight;
                final float dstAspect = (float) dstWidth / (float) dstHeight;
                if (srcAspect > dstAspect)
                    return new Rect(0, 0, dstWidth, (int) (dstWidth / srcAspect));
                else
                    return new Rect(0, 0, (int) (dstHeight * srcAspect), dstHeight);
            case CROP:
                return new Rect(0, 0, dstWidth, dstHeight);
            case ALL_TOP:
                return new Rect(0, 0, dstWidth, Math.min(dstHeight, srcHeight * dstWidth / srcWidth));
            default:
                return null;
        }
    }

    public Bitmap createScaledBitmap(Bitmap unscaledBitmap, int dstWidth, int dstHeight, ScalingLogic scalingLogic, Bitmap.Config config) {
        Rect srcRect = calculateSrcRect(unscaledBitmap.getWidth(), unscaledBitmap.getHeight(), dstWidth, dstHeight, scalingLogic);
        Rect dstRect = calculateDstRect(unscaledBitmap.getWidth(), unscaledBitmap.getHeight(), dstWidth, dstHeight, scalingLogic);
        Bitmap scaledBitmap = Bitmap.createBitmap(dstRect.width(), dstRect.height(), config);
        Canvas canvas = new Canvas(scaledBitmap);
        canvas.drawBitmap(unscaledBitmap, srcRect, dstRect, new Paint(Paint.FILTER_BITMAP_FLAG));
        return scaledBitmap;
    }

    public Bitmap decodeFile(int resourceId, int dstWidth, int dstHeight, ScalingLogic scalingLogic, Resources resources) {
        BitmapFactory.Options options = new BitmapFactory.Options();
        options.inScaled = false;
        return BitmapFactory.decodeResource(resources, resourceId, options);
    }

    public Bitmap decodeFile(String file, int dstWidth, int dstHeight, ScalingLogic scalingLogic, Resources resources) {
        BitmapFactory.Options options = new BitmapFactory.Options();
        options.inScaled = false;
        return BitmapFactory.decodeFile(file, options);
    }

    public Bitmap decodeInputStream(InputStream inputStream) {
        BitmapFactory.Options options = new BitmapFactory.Options();
        options.inScaled = false;
        return BitmapFactory.decodeStream(inputStream, null, options);
    }

    public void toGrayscale(ImageView imageView) {
        ColorMatrix matrix = new ColorMatrix();
        matrix.setSaturation(0);

        ColorMatrixColorFilter filter = new ColorMatrixColorFilter(matrix);
        imageView.setColorFilter(filter);
    }

    public void toNormalscale(ImageView imageView) {

        ColorMatrix matrix = new ColorMatrix();

        ColorMatrixColorFilter filter = new ColorMatrixColorFilter(matrix);
        imageView.setColorFilter(filter);
    }

}

File: app/src/main/java/xyz/lrhm/komakdast/core/util/legacy/ImageManager.kt
package xyz.lrhm.komakdast.core.util.legacy

import android.app.ActivityManager
import android.content.Context
import android.content.res.Resources
import android.graphics.*
import android.util.LruCache
import android.widget.ImageView
import dagger.hilt.android.qualifiers.ApplicationContext
import java.io.InputStream
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ImageManager @Inject constructor(
    @ApplicationContext val context: Context,
    private val lengthManager: LengthManager

) {

    var memoryClass = 0
    var cache: LruCache<ImageKey, Bitmap> = initCache(context)
    
    private fun initCache(context: Context): LruCache<ImageKey, Bitmap> {

        val am = context
            .getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
        memoryClass = am.memoryClass
        val max =
            (memoryClass * 1024 * 1024 * if (memoryClass > 100) 0.80 else 0.666).toInt() // more than 50%
        return object : LruCache<ImageKey, Bitmap>(max) {
            override fun sizeOf(key: ImageKey, value: Bitmap): Int {
                return value.byteCount
            }
        }
    }

    @JvmOverloads
    fun loadImageFromResource(
        resourceId: Int,
        outWidth: Int,
        outHeight: Int,
        scalingLogic: ScalingLogic? = ScalingLogic.CROP,
        config: Bitmap.Config? = Bitmap.Config.ARGB_8888
    ): Bitmap {
        var outWidth = outWidth
        var outHeight = outHeight
        if (outWidth == -1) outWidth = lengthManager.getWidthWithFixedHeight(resourceId, outHeight)
        if (outHeight == -1) outHeight = lengthManager.getHeightWithFixedWidth(resourceId, outWidth)
        val key = ImageKey(resourceId, outWidth, outHeight)
        val him = cache!![key]
        if (him != null && !him.isRecycled) return him
        System.gc()
        val scaledBitmap: Bitmap
        val unscaledBitmap: Bitmap
        unscaledBitmap =
            decodeFile(resourceId, outWidth, outHeight, scalingLogic, context.resources)
        scaledBitmap = createScaledBitmap(unscaledBitmap, outWidth, outHeight, scalingLogic, config)
        if (!unscaledBitmap.isRecycled) unscaledBitmap.recycle()
        cache!!.put(key, scaledBitmap)
        return scaledBitmap
    }

    fun loadImageFromFilse(
        path: String?,
        outWidth: Int,
        outHeight: Int,
        scalingLogic: ScalingLogic?
    ): Bitmap {
        val key = ImageKey(path!!, outWidth, outHeight)
        val him = cache!![key]
        if (him != null && !him.isRecycled) return him
        System.gc()
        val scaledBitmap: Bitmap
        val unscaledBitmap: Bitmap
        unscaledBitmap = decodeFile(path, outWidth, outHeight, scalingLogic, context.resources)
        scaledBitmap = createScaledBitmap(
            unscaledBitmap,
            outWidth,
            outHeight,
            scalingLogic,
            Bitmap.Config.ARGB_8888
        )
        if (!unscaledBitmap.isRecycled) unscaledBitmap.recycle()
        cache!!.put(key, scaledBitmap)
        return scaledBitmap
    }

    fun loadImageFromResourceNoCache(
        resourceId: Int,
        outWidth: Int,
        outHeight: Int,
        scalingLogic: ScalingLogic?,
        config: Bitmap.Config?
    ): Bitmap {
        return loadImageFromResource(resourceId, outWidth, outHeight, scalingLogic, config)
    }

    fun loadImageFromResourceNoCache(
        resourceId: Int,
        outWidth: Int,
        outHeight: Int,
        scalingLogic: ScalingLogic?
    ): Bitmap {
        return loadImageFromResource(
            resourceId,
            outWidth,
            outHeight,
            scalingLogic,
            Bitmap.Config.ARGB_8888
        )
    }

    fun loadImageFromResource(
        resourceId: Int,
        outWidth: Int,
        outHeight: Int,
        config: Bitmap.Config?
    ): Bitmap {
        return loadImageFromResource(resourceId, outWidth, outHeight, ScalingLogic.CROP, config)
    }

    fun loadImageFromInputStream(inputStream: InputStream?, outWidth: Int, outHeight: Int): Bitmap {
        var outWidth = outWidth
        var outHeight = outHeight
        System.gc()
        val scaledBitmap: Bitmap
        val unscaledBitmap: Bitmap?
        checkNotNull(inputStream) { "null InputStream!" }
        unscaledBitmap = decodeInputStream(inputStream)
        if (outHeight == -1) outHeight = unscaledBitmap!!.height * outWidth / unscaledBitmap.width
        if (outWidth == -1) outWidth = unscaledBitmap!!.width * outHeight / unscaledBitmap.height
        scaledBitmap = createScaledBitmap(
            unscaledBitmap,
            outWidth,
            outHeight,
            ScalingLogic.CROP,
            Bitmap.Config.ARGB_8888
        )
        if (!unscaledBitmap!!.isRecycled) unscaledBitmap.recycle()
        return scaledBitmap
    }

    enum class ScalingLogic {
        FIT, CROP, ALL_TOP
    }

    fun calculateSrcRect(
        srcWidth: Int,
        srcHeight: Int,
        dstWidth: Int,
        dstHeight: Int,
        scalingLogic: ScalingLogic?
    ): Rect? {
        return when (scalingLogic) {
            ScalingLogic.CROP -> {
                val srcAspect = srcWidth.toFloat() / srcHeight.toFloat()
                val dstAspect = dstWidth.toFloat() / dstHeight.toFloat()
                if (srcAspect > dstAspect) {
                    val srcRectWidth = (srcHeight * dstAspect).toInt()
                    val srcRectLeft = (srcWidth - srcRectWidth) / 2
                    Rect(srcRectLeft, 0, srcRectLeft + srcRectWidth, srcHeight)
                } else {
                    val srcRectHeight = (srcWidth / dstAspect).toInt()
                    val scrRectTop = (srcHeight - srcRectHeight) / 2
                    Rect(0, scrRectTop, srcWidth, scrRectTop + srcRectHeight)
                }
            }
            ScalingLogic.FIT -> Rect(0, 0, srcWidth, srcHeight)
            ScalingLogic.ALL_TOP -> Rect(
                0,
                0,
                srcWidth,
                Math.min(srcHeight, dstHeight * srcWidth / dstWidth)
            )
            else -> null
        }
    }

    fun calculateDstRect(
        srcWidth: Int,
        srcHeight: Int,
        dstWidth: Int,
        dstHeight: Int,
        scalingLogic: ScalingLogic?
    ): Rect? {
        return when (scalingLogic) {
            ScalingLogic.FIT -> {
                val srcAspect = srcWidth.toFloat() / srcHeight.toFloat()
                val dstAspect = dstWidth.toFloat() / dstHeight.toFloat()
                if (srcAspect > dstAspect) Rect(
                    0,
                    0,
                    dstWidth,
                    (dstWidth / srcAspect).toInt()
                ) else Rect(0, 0, (dstHeight * srcAspect).toInt(), dstHeight)
            }
            ScalingLogic.CROP -> Rect(0, 0, dstWidth, dstHeight)
            ScalingLogic.ALL_TOP -> Rect(
                0,
                0,
                dstWidth,
                Math.min(dstHeight, srcHeight * dstWidth / srcWidth)
            )
            else -> null
        }
    }

    fun createScaledBitmap(
        unscaledBitmap: Bitmap?,
        dstWidth: Int,
        dstHeight: Int,
        scalingLogic: ScalingLogic?,
        config: Bitmap.Config?
    ): Bitmap {
        val srcRect = calculateSrcRect(
            unscaledBitmap!!.width,
            unscaledBitmap.height,
            dstWidth,
            dstHeight,
            scalingLogic
        )
        val dstRect = calculateDstRect(
            unscaledBitmap.width,
            unscaledBitmap.height,
            dstWidth,
            dstHeight,
            scalingLogic
        )
        val scaledBitmap = Bitmap.createBitmap(dstRect!!.width(), dstRect.height(), config!!)
        val canvas = Canvas(scaledBitmap)
        canvas.drawBitmap(unscaledBitmap, srcRect, dstRect, Paint(Paint.FILTER_BITMAP_FLAG))
        return scaledBitmap
    }

    fun decodeFile(
        resourceId: Int,
        dstWidth: Int,
        dstHeight: Int,
        scalingLogic: ScalingLogic?,
        resources: Resources?
    ): Bitmap {
        val options = BitmapFactory.Options()
        options.inScaled = false
        return BitmapFactory.decodeResource(resources, resourceId, options)
    }

    fun decodeFile(
        file: String?,
        dstWidth: Int,
        dstHeight: Int,
        scalingLogic: ScalingLogic?,
        resources: Resources?
    ): Bitmap {
        val options = BitmapFactory.Options()
        options.inScaled = false
        return BitmapFactory.decodeFile(file, options)
    }

    fun decodeInputStream(inputStream: InputStream?): Bitmap? {
        val options = BitmapFactory.Options()
        options.inScaled = false
        return BitmapFactory.decodeStream(inputStream, null, options)
    }

    fun toGrayscale(imageView: ImageView) {
        val matrix = ColorMatrix()
        matrix.setSaturation(0f)
        val filter = ColorMatrixColorFilter(matrix)
        imageView.colorFilter = filter
    }

    fun toNormalscale(imageView: ImageView) {
        val matrix = ColorMatrix()
        val filter = ColorMatrixColorFilter(matrix)
        imageView.colorFilter = filter
    }


}